google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/QueryParameterValue.java

```diff
@@ -0,0 +1,433 @@
+
+  private static final DateTimeFormatter timestampFormatter =
+      DateTimeFormat.forPattern("yyyy-MM-dd HH:mm:ss.SSSSSSZZ").withZone(DateTimeZone.UTC);
+  private static final DateTimeFormatter dateFormatter = DateTimeFormat.forPattern("yyyy-MM-dd");
+  private static final DateTimeFormatter timeFormatter =
+      DateTimeFormat.forPattern("HH:mm:ss.SSSSSS");
+  private static final DateTimeFormatter datetimeFormatter =
+      DateTimeFormat.forPattern("yyyy-MM-dd HH:mm:ss.SSSSSS");
+
+  static final Function<
+          QueryParameterValue, com.google.api.services.bigquery.model.QueryParameterValue>
+      TO_VALUE_PB_FUNCTION =
+          new Function<
+              QueryParameterValue, com.google.api.services.bigquery.model.QueryParameterValue>() {
+            @Override
+            public com.google.api.services.bigquery.model.QueryParameterValue apply(
+                QueryParameterValue value) {
+              return value.toValuePb();
+            }
+          };
+  private static final long serialVersionUID = -5620695863123562896L;
+
+  private final String value;
+  private final StandardSQLTypeName type;
+  private final List<QueryParameterValue> arrayValues;
+  private final StandardSQLTypeName arrayType;
+
+  public static final class Builder {
+
+    private String value;
+    private List<QueryParameterValue> arrayValues;
+    private StandardSQLTypeName type;
+    private StandardSQLTypeName arrayType;
+
+    private Builder() {}
+
+    private Builder(QueryParameterValue queryParameterValue) {
+      this.value = queryParameterValue.value;
+      this.arrayValues =
+          queryParameterValue.arrayValues == null
+              ? null
+              : Lists.newArrayList(queryParameterValue.arrayValues);
+      this.type = queryParameterValue.type;
+      this.arrayType = queryParameterValue.arrayType;
+    }
+
+    /** Sets the value to the given scalar value. */
+    public Builder setValue(String value) {
+      this.value = value;
+      return this;
+    }
+
+    /** Sets array values. The type must set to ARRAY. */
+    public Builder setArrayValues(List<QueryParameterValue> arrayValues) {
+      this.arrayValues = arrayValues == null ? null : Lists.newArrayList(arrayValues);
+      return this;
+    }
+
+    /** Sets the parameter data type. */
+    public Builder setType(StandardSQLTypeName type) {
+      this.type = checkNotNull(type);
+      return this;
+    }
+
+    /** Sets the data type of the array elements. The type must set to ARRAY. */
+    public Builder setArrayType(StandardSQLTypeName arrayType) {
+      this.arrayType = arrayType;
+      return this;
+    }
+
+    /** Creates a {@code QueryParameterValue} object. */
+    public QueryParameterValue build() {
+      return new QueryParameterValue(this);
+    }
+  }
+
+  private QueryParameterValue(Builder builder) {
+    checkArgument(
+        (builder.arrayValues != null) == (StandardSQLTypeName.ARRAY.equals(builder.type)));
```

Comment:
I think it would have been more readable if it was written this way:
if (builder.arrayValues != null) {
   ... apply array precondition with relevant description of failure
} else {
  ... apply non array preconditions with relevant description of failure
}